File: app/src/test/java/be/robinj/distrohopper/AppManagerTest.java
package be.robinj.distrohopper;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import androidx.test.core.app.ActivityScenario;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.LooperMode;
import org.robolectric.shadows.ShadowLooper;

import java.util.List;

@RunWith(RobolectricTestRunner.class)
@LooperMode(LooperMode.Mode.LEGACY)
public class AppManagerTest {

    private ActivityScenario<HomeActivity> scenario;

    @Before
    public void setUp() {
        HomeActivityTest.seedPackageManager();
        scenario = ActivityScenario.launch(HomeActivity.class);
        Robolectric.flushBackgroundThreadScheduler();
        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();
    }

    @After
    public void tearDown() {
        scenario.close();
    }

    // ── App model ─────────────────────────────────────────────────────────────

    @Test
    public void everyAppHasNonEmptyPackageName() {
        scenario.onActivity(activity -> {
            for (App app : activity.getAppManager()) {
                assertFalse("Package name must not be empty for " + app.getLabel(),
                        app.getPackageName().isEmpty());
            }
        });
    }

    @Test
    public void everyAppHasNonEmptyLabel() {
        scenario.onActivity(activity -> {
            for (App app : activity.getAppManager()) {
                assertFalse("Label must not be empty for " + app.getPackageName(),
                        app.getLabel().isEmpty());
            }
        });
    }

    @Test
    public void getPackageAndActivityNameContainsBothParts() {
        scenario.onActivity(activity -> {
            App app = activity.getAppManager().get(0);
            String combined = app.getPackageAndActivityName();
            assertTrue(combined.contains(app.getPackageName()));
            assertTrue(combined.contains(app.getActivityName()));
            assertTrue("Must be newline-separated", combined.contains("\n"));
        });
    }

    @Test
    public void appEqualsItself() {
        scenario.onActivity(activity -> {
            App app = activity.getAppManager().get(0);
            assertEquals(app, app);
        });
    }

    @Test
    public void appsWithDifferentIdentitiesAreNotEqual() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            assertTrue("Need at least two apps", mgr.size() >= 2);
            App a = mgr.get(0);
            App b = mgr.get(1);
            if (!a.getPackageName().equals(b.getPackageName())
                    || !a.getActivityName().equals(b.getActivityName())) {
                assertNotEquals(a, b);
            }
        });
    }

    @Test
    public void appDoesNotEqualNonAppObject() {
        scenario.onActivity(activity -> {
            App app = activity.getAppManager().get(0);
            assertNotEquals(app, "a string");
            assertNotEquals(app, null);
        });
    }

    // ── Sorting ───────────────────────────────────────────────────────────────

    @Test
    public void installedAppsAreSortedAlphabetically() {
        scenario.onActivity(activity -> {
            List<App> apps = activity.getAppManager().getInstalledApps();
            assertTrue("Need at least two apps", apps.size() >= 2);
            for (int i = 0; i < apps.size() - 1; i++) {
                String a = apps.get(i).getLabel().toLowerCase();
                String b = apps.get(i + 1).getLabel().toLowerCase();
                assertTrue("Expected '" + a + "' <= '" + b + "'", a.compareTo(b) <= 0);
            }
        });
    }

    // ── Search ────────────────────────────────────────────────────────────────

    @Test
    public void searchEmptyPatternReturnsAllApps() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            assertEquals(mgr.size(), mgr.search("").size());
        });
    }

    @Test
    public void searchWithNoMatchReturnsEmpty() {
        scenario.onActivity(activity ->
                assertEquals(0, activity.getAppManager().search("xyzqvnomatch123").size()));
    }

    @Test
    public void searchIsCaseInsensitive() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            String label = mgr.get(0).getLabel();
            assertEquals(mgr.search(label.toLowerCase()).size(),
                         mgr.search(label.toUpperCase()).size());
        });
    }

    @Test
    public void searchMatchesByLabelPrefix() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            App first = mgr.get(0);
            String prefix = first.getLabel().substring(0, Math.min(3, first.getLabel().length()));
            assertTrue(mgr.search(prefix).contains(first));
        });
    }

    @Test
    public void searchResultsAreASubsetOfAllApps() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            List<App> all = mgr.getInstalledApps();
            for (App result : mgr.search("a")) {
                assertTrue("Search result must be in the full app list", all.contains(result));
            }
        });
    }

    @Test
    public void searchWithMaxResultsLimitsOutput() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            assertTrue(mgr.search("a", 2).size() <= 2);
        });
    }

    // ── Pinning ───────────────────────────────────────────────────────────────

    @Test
    public void pinningAnAppMarksItAsPinned() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            App app = findUnpinnedApp(mgr);
            assertNotNull("Need at least one unpinned app", app);
            assertFalse(mgr.isPinned(app));
            mgr.pin(app, false, false, false);
            assertTrue(mgr.isPinned(app));
        });
    }

    @Test
    public void pinningIncreasesThePinnedCount() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            App app = findUnpinnedApp(mgr);
            assertNotNull(app);
            int before = mgr.getPinned().size();
            mgr.pin(app, false, false, false);
            assertEquals(before + 1, mgr.getPinned().size());
        });
    }

    @Test
    public void pinningAlreadyPinnedAppIsIdempotent() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            App app = findUnpinnedApp(mgr);
            assertNotNull(app);
            mgr.pin(app, false, false, false);
            int countAfterFirst = mgr.getPinned().size();
            mgr.pin(app, false, false, false);
            assertEquals(countAfterFirst, mgr.getPinned().size());
        });
    }

    @Test
    public void findAppByPackageAndActivityNameReturnsCorrectApp() {
        scenario.onActivity(activity -> {
            AppManager mgr = activity.getAppManager();
            App expected = mgr.get(0);
            assertEquals(expected,
                    mgr.findAppByPackageAndActivityName(
                            expected.getPackageName(), expected.getActivityName()));
        });
    }

    @Test
    public void findAppByPackageAndActivityNameReturnsNullForUnknown() {
        scenario.onActivity(activity ->
                assertNull(activity.getAppManager()
                        .findAppByPackageAndActivityName("no.such.package", "NoSuchActivity")));
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static App findUnpinnedApp(AppManager mgr) {
        for (App app : mgr) {
            if (!mgr.isPinned(app)) return app;
        }
        return null;
    }
}

File: app/src/test/java/be/robinj/distrohopper/AppManagerTest.kt
package be.robinj.distrohopper

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import androidx.test.core.app.ActivityScenario
import androidx.test.core.app.ApplicationProvider
import be.robinj.distrohopper.R
import be.robinj.distrohopper.cache.AppIconCache
import be.robinj.distrohopper.preferences.Preference
import be.robinj.distrohopper.preferences.PreferencesActivity
import be.robinj.distrohopper.preferences.Preferences
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.LooperMode

@RunWith(RobolectricTestRunner::class)
@LooperMode(LooperMode.Mode.LEGACY)
class AppManagerTest {
    private lateinit var scenario: ActivityScenario<HomeActivity>

    @Before fun setUp() { scenario = ActivityTestSupport.launchHome() }
    @After fun tearDown() { scenario.close() }

    private fun withManager(block: (AppManager) -> Unit) {
        scenario.onActivity { block(it.appManager) }
    }
    private fun AppManager.unpinned() = firstOrNull { !isPinned(it) }
    private fun AppManager.settingsShortcut() = ActivityTestSupport.settingsShortcut(this.context)

    private fun cacheAnIcon(key: String): AppIconCache {
        val ctx = ApplicationProvider.getApplicationContext<android.content.Context>()
        val cache = AppIconCache(ctx)
        cache.put(key, BitmapDrawable(ctx.resources,
            Bitmap.createBitmap(8, 8, Bitmap.Config.ARGB_8888)))
        return cache
    }

    private fun cacheContains(key: String): Boolean =
        AppIconCache(ApplicationProvider.getApplicationContext()).containsKey(key)

    @Test fun changingTheIconConfigPurgesTheIconCache() = withManager { manager ->
        val key = "purge-me"
        assertTrue(cacheAnIcon(key).containsKey(key))

        // A new shape makes the icon config signature no longer match the cache. //
        Preferences.getSharedPreferences(manager.context).edit()
            .putString(Preference.ICON_SHAPE.getName(), "circle").commit()
        manager.iconRenderer // rebuilds the renderer, which reconciles the cache

        assertFalse(cacheContains(key))
    }

    @Test fun unchangedIconConfigKeepsTheIconCache() = withManager { manager ->
        val key = "keep-me"
        assertTrue(cacheAnIcon(key).containsKey(key))

        manager.iconRenderer // same config as at launch: nothing to reconcile

        assertTrue(cacheContains(key))
    }

    @Test fun everyAppHasNonEmptyPackageName() = withManager { manager ->
        manager.forEach { assertTrue(it.packageName.isNotEmpty()) }
    }

    @Test fun everyAppHasNonEmptyLabel() = withManager { manager ->
        manager.forEach { assertTrue(it.label.isNotEmpty()) }
    }

    @Test fun getPackageAndActivityNameContainsBothParts() = withManager {
        val app = it[0]
        assertEquals("${app.packageName}\n${app.activityName}", app.packageAndActivityName)
    }

    @Test fun appEqualsItself() = withManager { assertEquals(it[0], it[0]) }

    @Test fun appsWithDifferentIdentitiesAreNotEqual() = withManager { assertNotEquals(it[0], it[1]) }

    @Test fun appDoesNotEqualNonAppObject() = withManager {
        assertNotEquals(it[0], "a string")
        assertNotEquals(it[0], null)
    }

    @Test fun equalAppsHaveEqualHashCodes() = withManager { manager ->
        val app = manager[0]
        val sameIdentity = App(
            manager.context,
            manager,
            ActivityTestSupport.resolveInfo(app.packageName, app.activityName, app.label),
        )

        assertNotSame(app, sameIdentity)
        assertEquals(app, sameIdentity)
        assertEquals(app.hashCode(), sameIdentity.hashCode())
    }

    @Test fun installedAppsAreSortedAlphabetically() = withManager {
        assertEquals(it.installedApps.sortedBy { app -> app.label.lowercase() }, it.installedApps)
    }

    @Test fun searchEmptyPatternReturnsAllApps() = withManager { assertEquals(it.size(), it.search("").size) }

    @Test fun searchWithNoMatchReturnsEmpty() = withManager { assertTrue(it.search("xyzqvnomatch123").isEmpty()) }

    @Test fun searchIsCaseInsensitive() = withManager {
        assertEquals(it.search("alpha"), it.search("ALPHA"))
    }

    @Test fun searchMatchesByLabelPrefix() = withManager { assertTrue(it.search("Alp").contains(it[0])) }

    @Test fun searchResultsAreASubsetOfAllApps() = withManager { manager ->
        assertTrue(manager.installedApps.containsAll(manager.search("a")))
    }

    @Test fun searchWithMaxResultsLimitsOutput() = withManager { assertTrue(it.search("a", 2).size <= 2) }

    @Test fun searchFindsSettingsShortcutByPrefix() = withManager { manager ->
        assertEquals(listOf(manager.settingsShortcut()), manager.search("DistroHopper"))
    }

    @Test fun fullSearchFindsInfixMatches() = withManager { manager ->
        val settingsLabel = manager.context.getString(R.string.shortcut_label_distrohopper_settings)
        assertEquals(listOf(settingsLabel, "Settings"), manager.search("ting").map(App::getLabel))
    }

    @Test fun distroHopperSettingsShortcutReplacesOwnLauncherApp() = withManager { manager ->
        val packageName = ApplicationProvider.getApplicationContext<android.app.Application>().packageName

        assertNull(manager.findAppByPackageAndActivityName(packageName, HomeActivity::class.java.name))
        assertNotNull(manager.findAppByPackageAndActivityName(packageName, PreferencesActivity::class.java.name))
        assertTrue(manager.installedApps.any {
            it.label == manager.context.getString(R.string.shortcut_label_distrohopper_settings)
        })
    }

    @Test fun prefixOnlySearchRejectsInfixMatches() {
        scenario.onActivity { activity ->
            Preferences.getSharedPreferences(activity).edit()
                .putBoolean(Preference.DASH_SEARCH_FULL.getName(), false).commit()
            assertTrue(activity.appManager.search("ting").isEmpty())
        }
    }

    @Test fun pinningAnAppMarksItAsPinned() = withManager {
        val app = requireNotNull(it.unpinned()); assertTrue(it.pin(app, false, false)); assertTrue(it.isPinned(app))
    }

    @Test fun pinningIncreasesThePinnedCount() = withManager {
        val before = it.pinned.size; it.pin(requireNotNull(it.unpinned()), false, false); assertEquals(before + 1, it.pinned.size)
    }

    @Test fun pinningAlreadyPinnedAppIsIdempotent() = withManager {
        val app = requireNotNull(it.unpinned()); it.pin(app, false, false); assertFalse(it.pin(app, false, false))
    }

    @Test fun movePinnedAppChangesOrder() = withManager {
        val first = it[0]; val second = it[1]
        it.pin(first, false, false); it.pin(second, false, false); it.movePinnedApp(0, 1)
        assertEquals(listOf(second, first), it.pinned)
    }

    @Test fun settingsShortcutCanBePinnedMovedAndPersisted() = withManager { manager ->
        val alpha = manager.findAppsByPackageName("com.example.alpha").first()
        val settings = manager.settingsShortcut()

        manager.pin(alpha, false, false)
        manager.pin(settings, false, false)
        manager.movePinnedApp(1, 0)
        manager.savePinnedApps()

        assertEquals(listOf(settings, alpha), manager.pinned)
        val pinnedPrefs = Preferences.getSharedPreferences(manager.context, Preferences.PINNED_APPS)
        val desktops = PinnedAppsStorage.read(pinnedPrefs)
        assertEquals(
            listOf(settings.packageAndActivityName, alpha.packageAndActivityName), desktops[0])
    }

    @Test fun findAppByPackageAndActivityNameReturnsCorrectApp() = withManager {
        val expected = it[0]; assertEquals(expected, it.findAppByPackageAndActivityName(expected.packageName, expected.activityName))
    }

    @Test fun findAppByPackageAndActivityNameReturnsNullForUnknown() = withManager {
        assertNull(it.findAppByPackageAndActivityName("no.such.package", "NoSuchActivity"))
    }

    @Test fun findAppsByPackageNameReturnsOnlyMatchingApps() = withManager {
        val expected = it[0]; assertEquals(listOf(expected), it.findAppsByPackageName(expected.packageName))
    }

    @Test fun installedAppsMapUsesCombinedIdentity() = withManager {
        assertEquals(it.size(), it.installedAppsMap.size); assertSame(it[0], it.installedAppsMap[it[0].packageAndActivityName])
    }

    @Test fun installedAppsMapIncludesSettingsShortcutIdentity() = withManager { manager ->
        val settings = manager.settingsShortcut()
        assertSame(settings, manager.installedAppsMap[settings.packageAndActivityName])
    }

    @Test fun packageQueryCanBeFiltered() = withManager {
        assertEquals(1, it.queryInstalledApps("com.example.alpha").size)
    }

    @Test fun refreshDashSortOrderReordersForUsageBasedOrder() {
        scenario.onActivity { activity ->
            val manager = activity.appManager
            Preferences.getSharedPreferences(activity, Preferences.APP_USAGE).edit().clear().commit()
            // Make the last app in the current (alphabetical) order the most used //
            val last = manager.installedApps.last()
            AppUsageStats(activity).apply {
                recordLaunch(last.profileScopedKey)
                recordLaunch(last.profileScopedKey)
            }
            Preferences.getSharedPreferences(activity).edit()
                .putString(Preference.APP_SORT_ORDER.getName(), "most_used").commit()

            manager.refreshDashSortOrder()

            assertEquals(last, manager.installedApps.first())
        }
    }

    @Test fun refreshDashSortOrderIsANoOpForAlphabeticalOrder() {
        scenario.onActivity { activity ->
            val manager = activity.appManager
            Preferences.getSharedPreferences(activity, Preferences.APP_USAGE).edit().clear().commit()
            AppUsageStats(activity).recordLaunch(manager.installedApps.last().profileScopedKey)
            Preferences.getSharedPreferences(activity).edit()
                .putString(Preference.APP_SORT_ORDER.getName(), "alphabetical").commit()
            val before = manager.installedApps.toList()

            manager.refreshDashSortOrder()

            assertEquals(before, manager.installedApps)
            assertEquals(manager.installedApps.sortedBy { it.label.lowercase() }, manager.installedApps)
        }
    }
}
